### ISSUE TITLE
modelo SVC

### ISSUE THREAD

**Issue by VascoLopo** [VascoLopo/ajuda#4]
import spacy
from sentence_transformers import SentenceTransformer
from collections import defaultdict
import json
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import itertools
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score
import numpy as np
from sklearn.svm import SVC
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score, confusion_matrix
import matplotlib.pyplot as plt
# Load the data
with open(r'C:\Users\Lopo\Downloads\project-10-at-2023-04-19-17-02-b9209ecb.json', encoding='utf-8') as f:
    data = json.load(f)
# Load the Portuguese model
nlp = spacy.load("pt_core_news_lg")
# Define the stopwords set
wordz = {"de", "a", "o", "que", "e", "do", "ao", "rt", "da", "em", "um", "para", "é", "com", "uma",
             "os", "no", "se", "na", "por", "mais", "as", "dos", "como", "mas", "foi", "ao", "ele", "das",
             "tem", "à", "seu", "sua", "ou", "ser", "quando", "muito", "há", "nos", "já", "está", "eu",
             "também", "só", "pelo", "pela", "até", "isso", "ela", "entre", "era", "depois", "sem", "mesmo",
             "aos", "ter", "seus", "quem", "nas", "me", "esse", "eles", "estão", "você", "tinha", "foram",
             "essa", "num", "nem", "suas", "meu", "às", "minha", "têm", "numa", "pelos", "elas", "havia",
             "seja", "qual", "será", "nós", "tenho", "lhe", "deles", "essas", "esses", "pelas", "este", "fosse",
             "dele", "tu", "te", "vocês", "vos", "hei", "há", "havemos", "hão", "houve", "houvemos", "houveram",
             "houvera", "houvéramos", "haja", "hajamos", "hajam", "for", "is", "are", "this", "the", "and", "with",
             "to", "into", "on", "in", "of", "at", " . ", " , "}

# Define a SentenceTransformer model
#model = SentenceTransformer('sentence-transformers/paraphrase-xlm-r-multilingual-v1')
model = SentenceTransformer('xlm-r-100langs-bert-base-nli-mean-tokens')
embeddings = [[] for _ in range(len(data))]
labels = [[] for _ in range(len(data))]
pos_tags = [[] for _ in range(len(data))]
parse_trees = [[] for _ in range(len(data))]

for i, item in enumerate(data):
    id = i
    annotations = item['annotations']
    doc = nlp(item['data']['text'])
    tokens = [token.lemma_.lower() for token in doc if token.text.lower() not in wordz]
    preprocessed_sentence = " ".join(tokens)
    for annotation in annotations:
        result = annotation['result']
        if result:
            value = result[0]['value']
            choices = value.get('choices', [])
            if choices:
                choice = choices[0]
                embedding = model.encode(preprocessed_sentence)
                pos_tag = [[token.text, token.pos_] for token in doc]
                parse_tree = str(list(doc.sents)[0].root)
                embeddings[id].append(embedding)
                labels[id].append(choice)
                pos_tags[id].append(pos_tag)
                parse_trees[id].append(parse_tree)

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    print("ID:", i)
    print("Embeddings:", embeddings[i])
    print("Labels:", labels[i])
    print("POS tags:", pos_tags[i])
    print("Parse trees:", parse_trees[i])
    print("")
import pandas as pd

# Create an empty DataFrame with the desired columns
df = pd.DataFrame(columns=['ID', 'Embeddings', 'Labels', 'POS Tags', 'Parse Trees'])

# Iterate over the embeddings, labels, POS tags, and parse trees
for i in range(len(embeddings)):
    # Modify the 'labels' list so that any case where labels are both 'Populism' and 'Non populism' is counted only as 'Populism'
    if 'Non populism' in labels[i] and 'Populism' not in labels[i]:
        labels[i] = ['Non populism']
    elif 'Populism' in labels[i]:
        labels[i] = ['Populism']

    # Create a new row of data for each item
    row_data = {'ID': i, 'Embeddings': embeddings[i], 'Labels': labels[i], 'POS Tags': pos_tags[i],
                'Parse Trees': parse_trees[i]}
    # Append the row to the DataFrame
    df = df.append(row_data, ignore_index=True)

# Set the display option to show all columns
pd.set_option('display.max_columns', None)

# Print the DataFrame
print(df.head(10))

sizes = set()
for i in range(len(embeddings)):
    for embedding in embeddings[i]:
        size = len(embedding)
        sizes.add(size)
print(sizes)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(embeddings, labels, test_size=0.25, random_state=42)

# Reshape the data to 2D arrays
X_train = [np.concatenate(x) for x in X_train]
X_test = [np.concatenate(x) for x in X_test]
y_train = [y[0] for y in y_train]
y_test = [y[0] for y in y_test]
# Fit a logistic regression model to the training data
clf = LogisticRegression(max_iter=1000)
clf.fit(X_train, y_train)
# Define the SVM with an RBF kernel
svm = SVC(kernel='rbf', random_state=42)

# Define the hyperparameters to search over
param_grid = {'C': [0.1, 1, 10], 'gamma': [0.1, 1, 'scale']}

# Use GridSearchCV to search over the hyperparameters
grid_search = GridSearchCV(svm, param_grid, cv=5, verbose=2)

# Train the SVM using the best hyperparameters found by GridSearchCV
grid_search.fit(list(X_train), list(y_train))

# Predict on the test set and print the accuracy
predictions = grid_search.predict(list(X_train))
accuracy = accuracy_score(list(y_train), predictions)
print("Accuracy:", accuracy)
